### ISSUE TITLE
[skypeweb]  No support for libpurple/pidgin 3 (unreleased)

### ISSUE THREAD

**Issue by je-vv** [EionRobb/skype4pidgin#538]
When attempting to build skypeweb against libpurple/pidgin, 1st one needs to patch Makefile so that pkg-config uses "purple-3" instead of "purple".  But even though that, compiling whether the master repo, or the 1.3 latest release, fails at:

++++
In file included from skypeweb_connection.h:22:0,
                 from skypeweb_connection.c:19:
libskypeweb.h:83:22: fatal error: dnsquery.h: No such file or directory
 #include "dnsquery.h"
                      ^
compilation terminated.
++++

Understood libpurple/pidgin 3 is still unreleased, but I need to use 3+.  Is there a branch, or a way to build against libpurple/pidgin 3?

Thanks,

**Comment by EionRobb** [comment]
Looks like you've been having fun trying out Pidgin 3 and finding very few other plugins that work with it :) 

Rewriting the skypeweb plugin to work with purple3 is definitely on my to-do list, but there's not currently any way to compile it with purple3. 

The main issue is switching out the http calls to use the new ones in purple3

**Comment by je-vv** [comment]
Got it, :-(  Thanks!

**Comment by je-vv** [comment]
Closing then.

**Comment by EionRobb** [comment]
Let's keep this open so I don't forget :)

**Comment by EionRobb** [comment]
There we go, a 'small' commit to add purple3 support.

Would be keen for feedback :)

**Comment by vinpublic** [comment]
I got these:

In file included from /usr/src/skypeweb-git/skypeweb/skypeweb_connection.h:22:0,
                 from /usr/src/skypeweb-git/skypeweb/skypeweb_connection.c:19:
/usr/src/skypeweb-git/skypeweb/libskypeweb.h:84:18: fatal error: http.h: No such file or directory
 #include "http.h"
                  ^
compilation terminated.
make[2]: *** [CMakeFiles/skypeweb.dir/build.make:63: CMakeFiles/skypeweb.dir/skypeweb_connection.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /usr/src/skypeweb-git/skypeweb/skypeweb_contacts.h:22:0,
                 from /usr/src/skypeweb-git/skypeweb/skypeweb_contacts.c:20:
/usr/src/skypeweb-git/skypeweb/libskypeweb.h:84:18: fatal error: http.h: No such file or directory
 #include "http.h"
                  ^
compilation terminated.
make[2]: *** [CMakeFiles/skypeweb.dir/build.make:87: CMakeFiles/skypeweb.dir/skypeweb_contacts.c.o] Error 1
In file included from /usr/src/skypeweb-git/skypeweb/skypeweb_login.h:22:0,
                 from /usr/src/skypeweb-git/skypeweb/skypeweb_login.c:19:
/usr/src/skypeweb-git/skypeweb/libskypeweb.h:84:18: fatal error: http.h: No such file or directory
 #include "http.h"
                  ^
compilation terminated.
make[2]: *** [CMakeFiles/skypeweb.dir/build.make:111: CMakeFiles/skypeweb.dir/skypeweb_login.c.o] Error 1
In file included from /usr/src/skypeweb-git/skypeweb/skypeweb_messages.h:22:0,
                 from /usr/src/skypeweb-git/skypeweb/skypeweb_messages.c:19:
/usr/src/skypeweb-git/skypeweb/libskypeweb.h:84:18: fatal error: http.h: No such file or directory
 #include "http.h"
                  ^
compilation terminated.
make[2]: *** [CMakeFiles/skypeweb.dir/build.make:135: CMakeFiles/skypeweb.dir/skypeweb_messages.c.o] Error 1
In file included from /usr/src/skypeweb-git/skypeweb/skypeweb_util.h:19:0,
                 from /usr/src/skypeweb-git/skypeweb/skypeweb_util.c:19:
/usr/src/skypeweb-git/skypeweb/libskypeweb.h:84:18: fatal error: http.h: No such file or directory
 #include "http.h"
                  ^
compilation terminated.
make[2]: *** [CMakeFiles/skypeweb.dir/build.make:159: CMakeFiles/skypeweb.dir/skypeweb_util.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:68: CMakeFiles/skypeweb.dir/all] Error 2
make: *** [Makefile:150: all] Error 2

using purple 2 though.

**Comment by az143** [comment]
after git-pulling today (head is at commit 765e51) this no longer builds 
with libpurple 2.11.0-0+deb8u2 (ie. debian stable):

cc -fPIC -O2 -g -pipe -shared -o libskypeweb.so libskypeweb.c skypeweb_connection.c skypeweb_contacts.c skypeweb_login.c skypeweb_messages.c skypeweb_util.c purple2compat/http.c purple2compat/purple-socket.c   `pkg-config purple glib-2.0 json-glib-1.0 zlib --libs --cflags`   -Ipurple2compat -g -ggdb
skypeweb_contacts.c: In function 'skypeweb_got_vm_download_info':
skypeweb_contacts.c:262:40: error: 'PURPLE_XFER_TYPE_RECEIVE' undeclared (first use in this function)

looks like the version compat check for the renamed PURPLE_XFER_RECEIVE in purplecompat.h is misfiring.

**Comment by EionRobb** [comment]
@az143 whoops, good catch.  Commit 1ebb0e4 should sort that out

**Comment by vinpublic** [comment]
and now it doesn't register in pidgin as a plugin... :)

**Comment by vinpublic** [comment]
$ pidgin --debug showed this:
...
(22:11:09) plugins: /usr/lib/purple-2/libskypeweb.so is not usable because the 'purple_init_plugin' symbol could not be found.  Does the plugin call the PURPLE_INIT_PLUGIN() macro?

**Comment by EionRobb** [comment]
Well, not the smoothest rollout of a code overhaul :)
@vinpublic 6295b29369df73935d472c8d50415fd762dd3acb should sort that

**Comment by vinpublic** [comment]
nope:

(23:41:05) http: Performing new request 0x24da9c0 to (null).
(23:41:05) http: Invalid URL requested.
(23:41:05) http: Request 0x24da9c0 performed without success.
(23:41:05) g_log: skypeweb_string_get_chunk: assertion 'chunk_start' failed
(23:41:05) connection: Connection error on 0x24da120 (reason: 2 description: Failed getting MSPRequ cookie)
(23:41:05) GLib: g_hash_table_remove_internal: assertion 'hash_table != NULL' failed
(23:41:05) GLib: g_hash_table_lookup: assertion 'hash_table != NULL' failed
**
ERROR:/usr/src/skypeweb-git/skypeweb/purple2compat/http.c:1679:purple_http_connection_free: assertion failed: (gc_list != NULL)
Segmentation fault

**Comment by EionRobb** [comment]
Whew ok, 848f1d6 is working for me now.  I did a really good job of breaking it before :)

The benefits are that the message stream is a lot faster and there should be a lot less cpu usage as it shouldn't need to renegotiate ssl for each connection (i.e. keepalives are working now)

**Comment by vinpublic** [comment]
thanks. works smoothly now

**Comment by zod6** [comment]
> finch --debug
```
(19:55:29) plugins: Auto-loading plugin /usr/local/lib/purple-3/libskypeweb3.so
(19:55:29) plugins: Failed to load plugin /usr/local/lib/purple-3/libskypeweb3.so: Protocol type is not registered
....
```
version is 1.3-9-g6a05f49

**Comment by EionRobb** [comment]
@zod6 could you try with that commit?  I think that should get it working :)

**Comment by zod6** [comment]
yeah, it shows under protocols now!

**Comment by EionRobb** [comment]
Released 1.4 which includes purple-3 support :)

Thanks everyone for helping test and debug :)
